src/Stratis.Bitcoin/Base/BaseFeature.cs

```diff
@@ -182,22 +184,79 @@ public sealed class BaseFeature : FullNodeFeature
+
+            if (configLines.Any(c => c.Contains(BaseFeature.RewindFlag)))
+            {
+                for (int i = 0; i < configLines.Length; i++)
+                {
+                    if (configLines[i].Contains(BaseFeature.RewindFlag))
+                        configLines[i] = $"{BaseFeature.RewindFlag}={height}";
+                }
+
+                File.WriteAllLines(nodeSettings.ConfigurationFile, configLines);
+            }
+            else
+            {
+                using (StreamWriter sw = File.AppendText(nodeSettings.ConfigurationFile))
+                {
+                    sw.WriteLine($"{BaseFeature.RewindFlag}={height}");
+                };
+            }
+        }
+
         /// <inheritdoc />
         public override async Task InitializeAsync()
         {
             // TODO rewrite chain starting logic. Tips manager should be used.
 
             await this.StartChainAsync().ConfigureAwait(false);
 
+            ChainedHeader initializedAt = this.chainIndexer.Tip;
+
             if (this.provenBlockHeaderStore != null)
             {
                 // If we find at this point that proven header store is behind chain we can rewind chain (this will cause a ripple effect and rewind block store and consensus)
                 // This problem should go away once we implement a component to keep all tips up to date
                 // https://github.com/stratisproject/StratisBitcoinFullNode/issues/2503
-                ChainedHeader initializedAt = await this.provenBlockHeaderStore.InitializeAsync(this.chainIndexer.Tip);
-                this.chainIndexer.Initialize(initializedAt);
+                initializedAt = await this.provenBlockHeaderStore.InitializeAsync(this.chainIndexer.Tip);
             }
 
+            var rewindHeight = this.nodeSettings.ConfigReader.GetOrDefault<int>(BaseFeature.RewindFlag, -1);
+            if (rewindHeight >= 0)
+            {
+                if (rewindHeight < initializedAt.Height)
+                {
+                    // Ensure that we don't try to rewind further than the coin view is capable of doing.
+                    var utxoSet = ((dynamic)this.consensusRules).UtxoSet;
+                    var coinDatabase = ((dynamic)utxoSet).ICoindb;
+                    ((dynamic)coinDatabase).Initialize();
+                    int minRewindHeight = ((dynamic)coinDatabase).GetMinRewindHeight();
+                    ((dynamic)coinDatabase).Dispose();
+
+                    if (minRewindHeight == -1 || rewindHeight < minRewindHeight)
+                    {
+                        this.logger.LogWarning($"Can't rewind below block at height {minRewindHeight}. Rewind ignored.");
+                    }
+                    else
+                    {
+                        initializedAt = initializedAt.GetAncestor(rewindHeight);
+                        this.logger.LogInformation($"Rewinding to block at height {rewindHeight}.");
```

Comment:
I might be blind here... but where does the actual rewinding happen then?

Reply:
The `rewindAt` value causes a cascade of rewinds in dependent components.